### ISSUE TITLE
Thoughts on JWT

### ISSUE THREAD

**Issue by ijlyttle** [r-box/boxr#128]
Hi @nathancday,

As you know, this is taking a life of its own. I've done a little more reading at Box about JWT authentication and service accounts, and I am coming up with *an* idea for how this might all play. 

Because this idea involves some renamings and some changes in default behavior, I figure I should write them out before proposing anything in code.

The idea would start with renaming `box_auth_jwt()` to `box_auth_service()`, with a signature:

```r
box_auth_service <- function(token_file = NULL, user_id = NULL)
```

I'm proposing to change the argument from `config_file` to `token_file` because (in my mind) the word "token" evokes more the idea that this file should be kept securely.

Onto the default behavior:

```r
token_file <- token_file %||% Sys.getenv("BOX_TOKEN_FILE") %||% "~/.boxr-auth/token.json"

# read token file
config <- jsonlite::fromJSON(token_file)

# ref: https://developer.box.com/docs/construct-jwt-claim-manually#section-3-create-jwt-assertion
user_id <- user_id %||% config$enterpriseID
```

The default behavior would be to authorize using the service account. If the user wanted to authorize using their own account (the spooky scenario), they would have to do something extra:

```r
box_auth_service(user_id = box_user_id())
```

This brings us to our first helper:

```r
box_user_id <- function() {
  Sys.getenv("BOX_USER_ID")
}
```

At the risk of digressing, I would add some verbiage to the feedback from `box_auth()` to help the user set their `BOX_USER_ID` environment-variable at the same time as they would set `BOX_CLIENT_ID`, etc.

Our next helper is to help a service-account share a folder with an actual user, as described [here](https://developer.box.com/docs/service-account#section-storing-content-in-the-service-account-vs-the-user-account). The idea is that the folders and files could be stored with the service, then shared with individual users:

```r
box_dir_invite(dir_id = box_getwd(), user_id = box_user_id()) {

  # ref: https://developer.box.com/reference#collaboration-object

} 
```

The idea would be that you would:

- authenticate using `box_auth_service()`
- create your folder at Box, in the service file-system
- invite your *real* self to collaborate

**Edit to add**

This could also be used the other way (maybe):

- authenticate using `box_auth()`
- create your folder at Box, in the user file-system
- invite the service account to collaborate using `box_dir_invite()` with `user_id` as the `enterpriseID` from the token-file.

This is the extent of my crazy idea. I am happy to create a branch from the current PR to see if this can be made to work...

 Thoughts?

**Comment by nathancday** [comment]
I don't think this is crazy idea. I see this type of workflow as the long term path to a more secured JWT experience. 

Some points of clarification:

* The suggested behavior looks similar to what we are doing now in `box_auth_jwt()`, the meaningful change is to use the 'service-account-ID' right?

* Perhaps rename the arg to `account_id` (and docs) to reflect/imply a bigger umbrella that users and service accounts sit under.

* If the Box app still has the ability to "Act as existing users" (which would be required for the authorizing themselves) then we'd still have the same security concerns, right?

* I've always had the impression that inviting the service account as a collaborator (the second method you suggested) is the best practice. [This section](https://developer.box.com/docs/service-account#section-storing-content-in-the-service-account-vs-the-user-account) talks about document ownership and durability consequences of that.

Big picture:

This type of workflow feels like a significant change from the existing as-user workflow boxr has. Not that it is bad, but I can see this having edges we have yet to encounter. I would say 100% green light on branching and taking a shot at developing this behavior. I would also be 100% okay with leaving all things JWT until the next CRAN release too.

Related side note:

I got a new phone number last night and am in the queue for Box support to help me un-screw myself from being able to access my account (personal and boxr.tester) via 2-factor-auth. So that will be a massive sticking point for developing, sorry :(

**Comment by ijlyttle** [comment]
New phone - who this :) 

No worries, I can use my Box account for testing, for now, then switch things over when things are sorted out.

I guess I have in mind a way to be flexible to support as many workflows as we can. For example:

- If you set up your service app to act on behalf of any user, then you can provide your `account_id` (I like that idea!)

- If you set up your service app to only on behalf of the service account, don't provide an `account_id`.

- If you want your files to stay in the service-account filespace - invite users while auth'ed with the service account.

- If you want your files to stay in the user-account filespace - invite the service account while auth'ed with your user account.

Let me see what I can do...

**Comment by ijlyttle** [comment]
Should have closed with #105
